Public/Src/Engine/Scheduler/PipExecutor.cs

```diff
@@ -2851,21 +2851,12 @@ private static int GetCacheLevelForLocality(PublishedEntryRefLocality locality)
                 new List<(FileArtifact, FileMaterializationInfo)>(pip.Outputs.Length);
 
             // Outputs should be the same as what was in the metadata section.
-            int outputHashIndex = 0;
-            for (int i = 0; i < pip.Outputs.Length; i++)
+            for (int i = 0; i < metadata.StaticOutputHashes.Count; i++)
             {
-                FileArtifactWithAttributes attributedOutput = pip.Outputs[i];
-                if (!attributedOutput.CanBeReferencedOrCached())
-                {
-                    // Skipping non-cacheable outputs (note that StoreContentForProcess does the same).
-                    continue;
-                }
-
+                FileMaterializationInfo materializationInfo = metadata.StaticOutputHashes[i].Info.ToFileMaterializationInfo(pathTable);
+                FileArtifactWithAttributes attributedOutput = pip.Outputs.Where(o => AbsolutePath.Create(pathTable, metadata.StaticOutputHashes[i].AbsolutePath) == o.Path).Single();

```

Comment:
>pip.Outputs.Where(o => AbsolutePath.Create(pathTable, metadata.StaticOutputHashes[i].AbsolutePath) == o.Path).Single(); [](start = 62, length = 119)

The use of `Where` makes the loop quadratic. You may want to create a transient mapping from paths to output file artifacts before the loop, and thus you can do look-up cheaply.

Reply:
made the change, thanks.

Reply:
Can we have the dictionary be placed into ObjectPool? See [http://index/?leftProject=BuildXL.Utilities&leftSymbol=ejggwqg9vtu5&file=Pools.cs](http://index/?leftProject=BuildXL.Utilities&leftSymbol=ejggwqg9vtu5&file=Pools.cs)

---
In reply to: [301331607](https://github.com/microsoft/BuildXL/pull/552#discussion_r301331607) [](ancestors = 301331607)

Reply:
made the change.